src/test/java/com/puppycrawl/tools/checkstyle/internal/ImmutabilityTest.java

```diff
@@ -0,0 +1,444 @@
+        "com.puppycrawl.tools.checkstyle.checks.naming.ParameterNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.PatternVariableNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.RecordComponentNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.RecordTypeParameterNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.StaticVariableNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.naming.TypeNameCheck",
+        "com.puppycrawl.tools.checkstyle.checks.whitespace.TypecastParenPadCheck",
+        // Failing 'testClassesWithoutImmutableFieldsAreStateful'
+        "com.puppycrawl.tools.checkstyle.checks.whitespace.ParenPadCheck"
+    );
+
+    /**
+     * ArchCondition for immutable fields.
+     */
+    private static final ArchCondition<JavaField> BE_IMMUTABLE = new ImmutableFieldArchCondition();
+
+    /**
+     * Map fo module full name to module details.
+     */
+    private static final Map<String, ModuleDetails> MODULE_DETAILS_MAP =
+        XmlMetaReader.readAllModulesIncludingThirdPartyIfAny().stream()
+            .collect(Collectors.toMap(ModuleDetails::getFullQualifiedName,
+                                      moduleDetails -> moduleDetails));
+
+    /**
+     * Test to ensure that fields in util classes are immutable.
+     *
+     * @noinspection JUnitTestMethodWithNoAssertions
+     * @noinspectionreason JUnitTestMethodWithNoAssertions - asserts in callstack,
+     *     but not in this method
+     */
+    @Test
+    public void testUtilClassesImmutability() {
+        final JavaClasses utilClasses = new ClassFileImporter()
+            .withImportOption(ImportOption.Predefined.DO_NOT_INCLUDE_TESTS)
+            .importPackages("com.puppycrawl.tools.checkstyle.utils",
+                            "com.puppycrawl.tools.checkstyle.checks.javadoc.utils");
+
+        final ArchCondition<JavaField> beSuppressedField = new SuppressionArchCondition(
+            SUPPRESSED_FIELDS_IN_UTIL_CLASSES, "be suppressed");
+
+        final ArchRule fieldsShouldBeImmutable = fields()
+            .that()
+            .areDeclaredInClassesThat()
+            .haveSimpleNameEndingWith("Util")
+            .should(BE_IMMUTABLE)
+            .andShould()
+            .beFinal()
+            .andShould()
+            .beStatic()
+            .orShould(beSuppressedField);
+
+        fieldsShouldBeImmutable.check(utilClasses);
+    }
+
+    /**
+     * Test to ensure modules annotated with {@link StatelessCheck} contain immutable fields
+     * and modules with mutable fields are annotated with {@link FileStatefulCheck} or
+     * {@link GlobalStatefulCheck}.
+     *
+     * @noinspection JUnitTestMethodWithNoAssertions
+     * @noinspectionreason JUnitTestMethodWithNoAssertions - asserts in callstack,
+     *     but not in this method
+     */
+    @Test
+    public void testModuleImmutability() {
```

Comment:
@Vyom-Yadav is there some benefit to having three rules in one unit test? I think it is better to keep paradigm of 1 rule == 1 unit test, especially with complicated rules.

Reply:
I brought this up that we could have multiple tests that need mutability checks. I am ok with splitting things, but I don't think we should duplicate code.

Reply:
There should be very little/ no code duplication; conditions can be extracted to fields in the test class.

Reply:
Done.